src/libAtomVM/bitstring.c

```diff
@@ -158,6 +224,55 @@ bool bitstring_utf16_encode(avm_int_t c, uint8_t *buf, enum BitstringFlags bs_fl
     return true;
 }
 
+bool bitstring_utf16_decode(int32_t *c, uint8_t *buf, size_t len, enum BitstringFlags bs_flags, size_t *out_size)
+{
+    *c = 0;
+    if (len == 0) {
+        return false;
+    } else if (bs_flags & LittleEndianIntegerMask) {
+        if (len >= 4 && ((*(buf + 1) & 0xD8) == 0xD8) && ((*(buf + 3) & 0xDC) == 0xDC)) {
+            *c |= (*(buf + 1) & 0x03) << 18;
+            *c |= (*(buf + 0) & 0xFF) << 10;
+            *c |= (*(buf + 3) & 0x03) << 8;
+            *c |= (*(buf + 2) & 0xFF);
+            *c += 0x10000;
+            if (*c < 0 || *c > 0x10FFFF) {
+                return false;
+            }
+            *out_size = 4;
+            return true;
+        } else if (len >= 2) {
```

Comment:
This logic could be written differently, in a way that may be clearer, checking first UTF-16 word and failing if it's D8xx but length is only 2.

Reply:
I haven't addressed this one yet.

Reply:
In conversation @pguyot said this was okay to keep.